rfcs/0013-configurable-and-composable-storage.md

```diff
@@ -0,0 +1,600 @@
+
+E.g., the `StorageVec<TStorage>` (note here that every _compound storage type_ can store only types that implement `Storage`) will store slices of whatever the value of the contained `TStorage` is. Thus, its `Value` will simply be `type Value = [TStorage::Value]`.
+
+Similarly, the `StorageMap` will store `(K, V)` tuples, where `V` must be `Storage`. Thus, its `Value` will simply be `type Value = (K, V::Value)`.
+
+Once the `Value` is specified in the `Storage` implementation the compiler can type-check the RHS of `:=` against the `Value` expected by the LHS. It is worth noticing here that we have all the information needed, to, in a case of a type-mismatch, provide a detailed [expressive diagnostics](./0012-expressive-diagnostics.md) explaining the error.
+
+Also, every storage type will always have the same way of configuring, defined by the `Value` type. This guarantees intuitive and consistent usage. E.g., for the `StorageVec`, developers will know that they always need to provide a slice of the values expected by the contained storage type.
+
+The provided RHS expression that returns the type specified by `Value` must be const eval, as it is now.
+
+In the second step, the compiler needs to calculate the concrete storage slots in which to store the _configured with_ RHS values.
+
+This is where the `Storage::Config` associated type comes in play, together with the `const fn` function `internal_get_config()`.
+
+Before generating the slots, the compiler already knows:
+- the `StorageKey` at which a particular `storage` element will be stored. This key is calculated based on the element name and namespace, or it is given using the `in` keyword. This storage key is, as mentioned below, called the _self key_ of the storage element.
+- the const eval calculated RHS which has the type `Value`.
+
+The `const fn internal_get_config()` takes exactly those two parameters and provides a result of type `Self::Config`. Every `Self::Config` type is a combination of `StorageConfig` types defined as:
+
+```Sway
+pub struct StorageConfig<TValue>
+{
+    pub storage_key: StorageKey,
+    pub value: TValue,
+}
+```
+
+Essentially, the `StorageConfig<TValue>` tells the compiler at which `storage_key` the `value` should be stored. This information is const calculated by every storage type.
+
+E.g., the `StorageBox<T>` simply stores a value of the type `T` at its `self_key`. Therefore, its `Config` will be defined as `type Config = StorageConfig<T>`.
+
+E.g., the `StorageVec<TStorage>` must be able to store the vector's length and the configuration of all the contained elements. Thus, its `Config` can be specified as `type Config = (StorageConfig<u64>, [TStorage::Config])`. The first part of the tuple is the length, and the second the slice of all the `Config`s of the contained storage type.
+
+In general, the `Config` type is defined recursively as:
+
+```
+<Config> := StorageConfig<TValue>
+            | [<Config>]
+            | (<Config>, ...);
+```
```

Comment:
The mechanism is sound, but we need better naming than "config". It should be clearer that `StorageConfig` represents a single address initializer for consecutive slots and that `Config` is a whole, potentially composed, initializer.

Maybe `StorageInit` and `AddressInit`?

Reply:
Thanks for the proposal! I am not happy with the term `Config` either. While the `Value` as the name of the associated type is intuitive, `Config` does not carry such clarity.

Here we need to be cautios of the overall "initialization" vs "configuration" confusion. Since `configurable`s and `storage` are actually _configured_ in code and that configuration can be changed at deployment time, the term "configuration" is consistently used everywhere in the RFC. Sticking to that term will help to remove existing confusion, where the choice of the syntax and wording implies "initialization" and "assignment" to something that is rather a "configuration".

E.g., the `Storage::init` constructor on the other side, is an "initialization" because it will create a storge type instance at the runtime and actually stores the provided values into the storage.

To the origin of the current, non-optimal name `Config`, which I tried to change (my thinking below). Since both the associated type `Config` and the `StorageConfig` are used only by the compiler during the configuration, I've kept the name `Config` although I don't like it.

Alternative names I've played with for the `Config` associated type were:
- `ValueLayout`: because it provides the information on how the value (described by the `Value` type) is layed out in the storage. The reason I didn't choose it is, that storage types can store additional information as well, e.g., `StorageVec` its length, and not only what is in the `Value`.
- `StorageLayout`: it clashed with the `StorageLayout` enum.
- `StorageContent`: because it describes which content is stored in the storage and where exactly.

Using "Init" in the name, as explained about, can lead back to the current confusion.

Perhaps `SlotsConfig` because it provides the information on how to configure the slots, or even better `SlotsContent`?

Additional proposals are welcome :smile:

Reply:
Also, regardless of the terminology we chose at the end, I like the proposed formulation (but to change "initializing" to "configuring"):

> StorageConfig represents a single address initializer for consecutive slots and that Config is a whole, potentially composed, initializer

Hmmm, perhaps:

`Storage(Address)Config/Content` represents a configuration for storing a single `value/content` in consecutive slots starting at the address specified by the `storage_key`. The `SlotsConfig/Content` is a whole, potentially composed, configuration, for storing the entire value contained in the `Storage`.

Whatever the final terminology will be, it should sound well in that sentence :smile: